### DISCUSSION TITLE
Plugin Support

### DISCUSSION THREAD

**Discussion by sammy-ette** [sammy-ette/Hilbish#25] [Ideas]
1st party plugin support for Hilbish
They would be called petals because Hibiscus and flowers :hibiscus:
So Hilbish would basically have a plugin manager builtin to help with updating, installing and loading them
Instead of spamming `require`s for specific paths, you would just `petals.load 'someone/hilbishplugin'` maybe like vim plugins

**Comment by devins2518** [comment]
Most plugin managers (at least vim-plug which seems to be the most popular) has support for local plugins too so we'll need to handle whether its a git directory or local folder. Easiest way that I can think of is to just assume its local every time and check if its a local folder otherwise try github.

**Reply by sammy-ette** [reply]
One thing we should do first is figure out how the file structure of a plugin should be
Manifest file, entry point, etc.

**Reply by devins2518** [reply]
Maybe something like
```
├── readme and other general info
├── require.lua (for dependencies, don't know if this will cause lua to freak out because of recursive naming)
├── lua
     ├── dir_with_cool_stuff    
     └── plugin.lua (actual plugin file that declares plugin name and other stuffs)

```
I'm not too well versed in lua dependency stuff like what awesome does for it, but this could be a good solution

**Reply by sammy-ette** [reply]
I don't think having the `require.lua` is a good idea, and also for the reason that you said, there should instead be 1 file to define the name, version, dependencies and other stuff about the plugin with a different filename.
`plugin.lua` will stay, but the folder name should be something different

**Reply by devins2518** [reply]
Hm, okay so maybe to load the plugin, only read from `lua/plugin.lua`. `plugin.lua` will contain the name, version, other plugin stuff, and just export whatever the public api is going to be.

**Reply by sammy-ette** [reply]
I meant more like how node.js/npm has packages
Another thing, how will a plugin for custom commands vs a traditional module be handled

**Reply by devins2518** [reply]
Hm so maybe like:
```
├── readme and other general info
├── package.lua (name declaration, description, version, etc etc)
├── lua
     ├── dir_with_cool_stuff
     ├── my_file.lua
     └── my_secret_file.lua
```
The name of `package.lua` can change as long as its the only lua file in the root directory. You could then load modules by `require package.my_file` or something like that. `package.lua` should be able to decide privacy somehow (maybe through a table of importable libs?). So if the table only included `my_file.lua` and you tried to do `require package.my_secret_file` it would just error out.

**Reply by sammy-ette** [reply]
The privacy thing seems like a very specific case that shouldnt have to be prioritized now

**Reply by devins2518** [reply]
Sure, but is the general layout looking sane enough

**Reply by sammy-ette** [reply]
So for plugin location how about `~/.local/share/petals/start` or something like that, it should be in `~/.local/share`

**Reply by devins2518** [reply]
Maybe `~/.local/share/hilbish/petals/start`

**Reply by sammy-ette** [reply]
Yes that's good

**Comment by sammy-ette** [comment]
Next thing is the petals module api, should it actually be `petals.load`? How will plugins be *loaded* if they arent installed, and if they are how will they be *started*

**Reply by sammy-ette** [reply]
figured that all out
check out petals at https://github.com/Hilbis/Petals

**Comment by sammy-ette** [comment]
how about just directly going the [packer](https://github.com/wbthomason/packer.nvim) approach to how it works? petals and packer are both lua
in the case of the neovim commands, those can just be interactive commands (update, install, which was the original plan)
